Src/GoogleApis.Tests/Apis/Requests/PageStreamerTest.cs

```diff
@@ -0,0 +1,241 @@
+                resource.NewRequest(),
+                CancellationToken.None);
+            Assert.AreEqual(resource.AllItems, actual);
+        }
+
+        // Note: this doesn't test all the possible timings of cancellation
+        [Test]
+        public async Task Cancellation()
+        {
+            var cts = new CancellationTokenSource();
+            var gatekeeperSource = new TaskCompletionSource<int>();
+            var resource = new PageStreamedResource(
+                "check",
+                new Page("x", 1, 2, 3),
+                new Page("y", 4, 5),
+                new Page(null, 6, 7));
+            resource.gateKeeper = gatekeeperSource.Task;            
+
+            var task = PageStreamedResource.PageStreamer.FetchAllAsync(resource.NewRequest(), cts.Token);
+            // Delay the resource fetching until we've cancelled the token
+            cts.Cancel();
+            gatekeeperSource.SetResult(0);
+            // TODO: Move to Assert.ThrowsAsync when we update to an appropriate version of NUnit.
+            // Assert.That is available, but the documentation is unclear how this interacts with await.
+            try
+            {
+                await task;
+                Assert.Fail("Expected exception");
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected
+            }
+        }
+
+        // Has to be public so we can use it as a parameter for test cases
+        public class PageStreamedResource
+        {
+            internal static readonly PageStreamer<int, Request, Page, string> PageStreamer =
+                new PageStreamer<int, Request, Page, string>(
+                    (request, token) => request.WithToken(token),
+                    page => page.NextPageToken,
+                    page => page.Items);
+
+            // Task to await when fetching pages; defaults to an already completed task,
+            // but can be set to another task to allow for more fine-grained interaction.
+            internal Task gateKeeper = TaskEx.FromResult(0);
+            internal List<Page> Pages { get; private set; }
+            internal string Name { get; set; }
+            internal string RequestCheck { get; set; }
+
+            internal PageStreamedResource(string name, params Page[] pages)
+            {                
+                this.Name = name;
+                this.Pages = pages.ToList();
+            }
+
+            public override string ToString()
+            {
+                return Name;
+            }
+
+            internal Request NewRequest()
+            {
+                return new Request(this) { Check = RequestCheck };
+            }
+
+            internal Page GetPage(string token, string check)
+            {
+                Assert.AreEqual(RequestCheck, check);
+                int index = Pages.FindIndex(page => page.NextPageToken == token);
```

Comment:
I think that something here confused me for long time. If we had two fields on page one for token and one for next token, it would make it easier to understand. Because right now, for example, I'm not sure why x, and y in simpleResource are important.
I believe that if it was hard for me to understand (and I'm still not sure that I got it in 100%), it might be hard for future developers who will try to add\modify something here.

You can remember how complicated it was to change something in MediaDownloader\Uploader tests (and you can blame me on that...), so maybe there is a room here to make it simpler.

On the other hand, Matt didn't complain about it, so maybe it's only me...